### ISSUE TITLE
How to publish during execution of RPC function?

### ISSUE THREAD

**Issue by vokson** [voryx/Thruway#188]
Hello!
I'm sorry to disturb you, but I can't understand how to implement problem mentioned above. My purpose to make PHP/Thruway client -> register function, then call this function from Browser/Autobahn.JS. I want to publish messages with information about execution percentage from PHP client **during execution** of function.
For the purpose I made class Pusher, which can store remaining messages, ans 0.5sec loop, which pulls message from the class and publish them.
My problem is: **When PHP client receives call, process finish callable function first and only after that goes into loop.**
Please, help how to manage the problem.

```php
use Thruway\ClientSession;
use Thruway\Peer\Client;
use Thruway\Transport\PawlTransportProvider;
use App\Classes\Message\Pusher;

$loop = \React\EventLoop\Factory::create();

$client = new Client('realm1', $loop);
$client->addTransportProvider(new PawlTransportProvider('ws://127.0.0.1:8080'));

$loop->addPeriodicTimer(0.5, function () use ($client, &$loop) {

    $session = $client->getSession();

    if ($session && ($session->getState() == ClientSession::STATE_UP)) {
        while ($push = Pusher::pull()) {
            $session->publish($push['channel'], [$push['message']]);
        }
    }
});

$client->on('open', function (ClientSession $session) {

    $add2= function ($args) {
        for ($i=0; $i<10; $i++) {
            sleep(1);
            Pusher::push('com.myapp.hello', uniqid());
        }

        return $args[0] + $args[1];
    };
    $session->register('com.myapp.add2', $add2);

});

$client->start();
```

```php
namespace App\Classes\Message;

class Pusher {
    static private $array = [];

    static public function push($channel, $message)
    {
        array_push(self::$array, ['channel' => $channel, 'message' => $message]);
    }

    static public function pull()
    {
        return array_shift(self::$array);
    }
}
````

**Comment by ledazinha** [comment]
You can't use sleep() within your wamp peer, because it will block the event loop, asynchronous I/O, and all timers.

In your case, your rpc callback enters a for() loop that basically stops program execution for 10 seconds. During the sleep the event loop doesn't run, and the timers don't work. After the sleep is finished, your callback immediately returns the result of the rpc call, and only then the timers will run again, hence your autobahn client will receive the result of the call first, and only after that, the publications.

I assume you are using sleep() to simulate an assynchonous operation that can take some time to complete. You can acchieve this using Promises and Observables. This is an example of how to use Promises to accomplish what you're trying to do:

```php
$add2 = function ($args) use ($loop) {
    $deferred = new \React\Promise\Deferred();
    $interval = 1;
    $counter = 0;
    
    $loop->addPeriodicTimer(
        $interval,
        function (\React\EventLoop\Timer\TimerInterface $timer) use ($args, $deferred, &$counter) {
            if ($counter < 10) {
                Pusher::push('com.myapp.hello', uniqid());
            } else {
                $timer->cancel();
                $deferred->resolve($args[0] + $args[1]); // this will be the result of your rpc call
            }
            $counter++;
        }
    );

    return $deferred->promise();
};

$session->register('com.myapp.add2', $add2);
```
NOTE¹: I haven't tested this code, so you might need to fiddle a little with it to make it work ;-)
NOTE²: This is just an example, and not the most efficient way to do this, but I wanted to keep it simple :-/

**Comment by mbonneau** [comment]
@vokson I am closing this issue - please reopen if you are still having issues.
